File: backend/app/api/api_v1/endpoints/tasks.py
"""
Task management endpoints
"""

from typing import List, Optional
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session

from app.core.database import get_db
from app.core.auth import get_current_user
from app.models.user import User
from app.schemas.task import (
    TaskCreate, TaskUpdate, TaskResponse, TaskListItem, TaskStatusUpdate,
    TaskBatchUpdate, TaskBatchStatusUpdate, TaskGenerateRequest, TaskGenerateResponse,
    TaskSearchRequest, TaskStats, TaskStatus, TaskLogResponse, TaskWithLogs
)
from app.services.task import task_service

router = APIRouter()


@router.get("/", response_model=List[TaskListItem])
async def get_tasks(
    project_id: Optional[int] = Query(None, description="项目ID筛选"),
    parent_id: Optional[int] = Query(None, description="父任务ID筛选"),
    status: Optional[List[TaskStatus]] = Query(None, description="状态筛选"),
    skip: int = Query(0, ge=0, description="跳过数量"),
    limit: int = Query(50, ge=1, le=100, description="限制数量"),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get tasks with optional filtering"""
    try:
        tasks = task_service.get_tasks(
            db=db,
            user_id=current_user.id,
            project_id=project_id,
            parent_id=parent_id,
            skip=skip,
            limit=limit
        )
        
        # Convert to list items with counts
        task_items = []
        for task in tasks:
            item_dict = {
                "id": task.id,
                "title": task.title,
                "description": task.description,
                "status": task.status,
                "priority": task.priority,
                "project_id": task.project_id,
                "parent_id": task.parent_id,
                "order_index": task.order_index,
                "estimated_hours": task.estimated_hours,
                "actual_hours": task.actual_hours,
                "assignee_id": task.assignee_id,
                "due_date": task.due_date,
                "completed_at": task.completed_at,
                "created_at": task.created_at,
                "updated_at": task.updated_at,
                "subtask_count": len(task.subtasks),
                "dependency_count": len(task.dependencies)
            }
            task_items.append(TaskListItem(**item_dict))
        
        return task_items
        
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/", response_model=TaskResponse)
async def create_task(
    task_data: TaskCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Create a new task"""
    try:
        task = task_service.create_task(db=db, task_data=task_data, user_id=current_user.id)
        return TaskResponse.from_orm(task)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/generate", response_model=TaskGenerateResponse)
async def generate_tasks(
    project_id: int,
    generate_request: TaskGenerateRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Generate tasks using AI"""
    try:
        result = task_service.generate_tasks_with_ai(
            db=db,
            user_id=current_user.id,
            project_id=project_id,
            generate_request=generate_request
        )
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/search", response_model=List[TaskListItem])
async def search_tasks(
    search_request: TaskSearchRequest = Depends(),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Search tasks with advanced filters"""
    try:
        tasks, total_count = task_service.search_tasks(
            db=db,
            user_id=current_user.id,
            search_request=search_request
        )
        
        # Convert to list items
        task_items = []
        for task in tasks:
            item_dict = {
                "id": task.id,
                "title": task.title,
                "description": task.description,
                "status": task.status,
                "priority": task.priority,
                "project_id": task.project_id,
                "parent_id": task.parent_id,
                "order_index": task.order_index,
                "estimated_hours": task.estimated_hours,
                "actual_hours": task.actual_hours,
                "assignee_id": task.assignee_id,
                "due_date": task.due_date,
                "completed_at": task.completed_at,
                "created_at": task.created_at,
                "updated_at": task.updated_at,
                "subtask_count": len(task.subtasks),
                "dependency_count": len(task.dependencies)
            }
            task_items.append(TaskListItem(**item_dict))
        
        return task_items
        
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/stats", response_model=TaskStats)
async def get_task_stats(
    project_id: Optional[int] = Query(None, description="项目ID筛选"),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get task statistics"""
    try:
        stats = task_service.get_task_stats(
            db=db,
            user_id=current_user.id,
            project_id=project_id
        )
        return stats
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/{task_id}", response_model=TaskResponse)
async def get_task(
    task_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get task by ID"""
    task = task_service.get_task(db=db, task_id=task_id, user_id=current_user.id)
    if not task:
        raise HTTPException(status_code=404, detail="Task not found")
    
    return TaskResponse.from_orm(task)


@router.put("/{task_id}", response_model=TaskResponse)
async def update_task(
    task_id: int,
    task_data: TaskUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Update task"""
    try:
        task = task_service.update_task(
            db=db,
            task_id=task_id,
            user_id=current_user.id,
            task_data=task_data
        )
        if not task:
            raise HTTPException(status_code=404, detail="Task not found")
        
        return TaskResponse.from_orm(task)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.patch("/{task_id}/status", response_model=TaskResponse)
async def update_task_status(
    task_id: int,
    status_data: TaskStatusUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Update task status"""
    try:
        task = task_service.update_task_status(
            db=db,
            task_id=task_id,
            user_id=current_user.id,
            status=status_data.status
        )
        if not task:
            raise HTTPException(status_code=404, detail="Task not found")
        
        return TaskResponse.from_orm(task)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.delete("/{task_id}")
async def delete_task(
    task_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Delete task"""
    success = task_service.delete_task(db=db, task_id=task_id, user_id=current_user.id)
    if not success:
        raise HTTPException(status_code=404, detail="Task not found")
    
    return {"message": "Task deleted successfully"}


@router.post("/batch/update", response_model=List[TaskResponse])
async def batch_update_tasks(
    batch_data: TaskBatchUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Batch update multiple tasks"""
    try:
        tasks = task_service.batch_update_tasks(
            db=db,
            user_id=current_user.id,
            batch_data=batch_data
        )
        return [TaskResponse.from_orm(task) for task in tasks]
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/batch/status", response_model=List[TaskResponse])
async def batch_update_status(
    batch_data: TaskBatchStatusUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Batch update task status"""
    try:
        tasks = task_service.batch_update_status(
            db=db,
            user_id=current_user.id,
            batch_data=batch_data
        )
        return [TaskResponse.from_orm(task) for task in tasks]
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/{task_id}/dependencies/{depends_on_id}")
async def add_dependency(
    task_id: int,
    depends_on_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Add a dependency between tasks"""
    try:
        success = task_service.add_dependency(
            db=db,
            task_id=task_id,
            depends_on_id=depends_on_id,
            user_id=current_user.id
        )
        if not success:
            raise HTTPException(status_code=404, detail="Task not found")
        
        return {"message": "Dependency added successfully"}
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.delete("/{task_id}/dependencies/{depends_on_id}")
async def remove_dependency(
    task_id: int,
    depends_on_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Remove a dependency between tasks"""
    try:
        success = task_service.remove_dependency(
            db=db,
            task_id=task_id,
            depends_on_id=depends_on_id,
            user_id=current_user.id
        )
        if not success:
            raise HTTPException(status_code=404, detail="Dependency not found")
        
        return {"message": "Dependency removed successfully"}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/{task_id}/logs", response_model=List[TaskLogResponse])
async def get_task_logs(
    task_id: int,
    skip: int = Query(0, ge=0, description="跳过数量"),
    limit: int = Query(50, ge=1, le=100, description="限制数量"),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get task change logs"""
    try:
        logs = task_service.get_task_logs(
            db=db,
            task_id=task_id,
            user_id=current_user.id,
            skip=skip,
            limit=limit
        )
        return [TaskLogResponse.from_orm(log) for log in logs]
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/{task_id}/with-logs", response_model=TaskWithLogs)
async def get_task_with_logs(
    task_id: int,
    logs_limit: int = Query(20, ge=1, le=100, description="日志数量限制"),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get task with its change logs"""
    task = task_service.get_task(db=db, task_id=task_id, user_id=current_user.id)
    if not task:
        raise HTTPException(status_code=404, detail="Task not found")
    
    try:
        logs = task_service.get_task_logs(
            db=db,
            task_id=task_id,
            user_id=current_user.id,
            skip=0,
            limit=logs_limit
        )
        
        task_dict = TaskResponse.from_orm(task).dict()
        task_dict["logs"] = [TaskLogResponse.from_orm(log) for log in logs]
        
        return TaskWithLogs(**task_dict)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

File: frontend/src/services/tasks.ts
/**
 * Tasks API services
 */

import api from '@/lib/api';
import {
  Task,
  TaskListItem,
  TaskCreate,
  TaskUpdate,
  TaskStats,
  TaskSearchRequest,
  TaskGenerateRequest,
  TaskGenerateResponse,
  TaskLog,
  TaskWithLogs,
  TaskBatchUpdate,
  TaskBatchStatusUpdate,
  TaskStatus,
} from '@/types/task';

export const tasksService = {
  // Get all tasks (global)
  async getTasks(params?: {
    project_id?: number;
    parent_id?: number;
    status?: TaskStatus[];
    skip?: number;
    limit?: number;
  }): Promise<TaskListItem[]> {
    const searchParams = new URLSearchParams();
    
    if (params?.project_id !== undefined) searchParams.append('project_id', params.project_id.toString());
    if (params?.parent_id !== undefined) searchParams.append('parent_id', params.parent_id.toString());
    if (params?.status) params.status.forEach(s => searchParams.append('status', s));
    if (params?.skip !== undefined) searchParams.append('skip', params.skip.toString());
    if (params?.limit !== undefined) searchParams.append('limit', params.limit.toString());
    
    const response = await api.get(`/tasks?${searchParams.toString()}`);
    return response.data;
  },

  // Get tasks for a specific project
  async getProjectTasks(projectId: number, params?: {
    parent_id?: number;
    skip?: number;
    limit?: number;
  }): Promise<TaskListItem[]> {
    const searchParams = new URLSearchParams();
    
    if (params?.parent_id !== undefined) searchParams.append('parent_id', params.parent_id.toString());
    if (params?.skip !== undefined) searchParams.append('skip', params.skip.toString());
    if (params?.limit !== undefined) searchParams.append('limit', params.limit.toString());
    
    const response = await api.get(`/projects/${projectId}/tasks?${searchParams.toString()}`);
    return response.data;
  },

  // Get a specific task by ID
  async getTask(id: number): Promise<Task> {
    const response = await api.get(`/tasks/${id}`);
    return response.data;
  },

  // Get task with logs
  async getTaskWithLogs(id: number, logsLimit: number = 20): Promise<TaskWithLogs> {
    const response = await api.get(`/tasks/${id}/with-logs?logs_limit=${logsLimit}`);
    return response.data;
  },

  // Create a new task
  async createTask(data: TaskCreate): Promise<Task> {
    const response = await api.post('/tasks', data);
    return response.data;
  },

  // Create a task for a specific project
  async createProjectTask(projectId: number, data: TaskCreate): Promise<Task> {
    const response = await api.post(`/projects/${projectId}/tasks`, data);
    return response.data;
  },

  // Update an existing task
  async updateTask(id: number, data: TaskUpdate): Promise<Task> {
    const response = await api.put(`/tasks/${id}`, data);
    return response.data;
  },

  // Update task status
  async updateTaskStatus(id: number, status: TaskStatus): Promise<Task> {
    const response = await api.patch(`/tasks/${id}/status`, { status });
    return response.data;
  },

  // Delete a task
  async deleteTask(id: number): Promise<void> {
    await api.delete(`/tasks/${id}`);
  },

  // Generate tasks using AI
  async generateTasks(projectId: number, request: TaskGenerateRequest): Promise<TaskGenerateResponse> {
    const response = await api.post(`/projects/${projectId}/tasks/generate`, request);
    return response.data;
  },

  // Search tasks
  async searchTasks(request: TaskSearchRequest): Promise<TaskListItem[]> {
    const searchParams = new URLSearchParams();
    
    if (request.query) searchParams.append('query', request.query);
    if (request.status) request.status.forEach(s => searchParams.append('status', s));
    if (request.priority) request.priority.forEach(p => searchParams.append('priority', p));
    if (request.assignee_id !== undefined) searchParams.append('assignee_id', request.assignee_id.toString());
    if (request.parent_id !== undefined) searchParams.append('parent_id', request.parent_id.toString());
    if (request.has_subtasks !== undefined) searchParams.append('has_subtasks', request.has_subtasks.toString());
    if (request.has_dependencies !== undefined) searchParams.append('has_dependencies', request.has_dependencies.toString());
    if (request.due_date_from) searchParams.append('due_date_from', request.due_date_from);
    if (request.due_date_to) searchParams.append('due_date_to', request.due_date_to);
    if (request.created_from) searchParams.append('created_from', request.created_from);
    if (request.created_to) searchParams.append('created_to', request.created_to);
    if (request.skip !== undefined) searchParams.append('skip', request.skip.toString());
    if (request.limit !== undefined) searchParams.append('limit', request.limit.toString());
    if (request.sort_by) searchParams.append('sort_by', request.sort_by);
    if (request.sort_order) searchParams.append('sort_order', request.sort_order);
    
    const response = await api.get(`/tasks/search?${searchParams.toString()}`);
    return response.data;
  },

  // Search tasks within a project
  async searchProjectTasks(projectId: number, request: TaskSearchRequest): Promise<TaskListItem[]> {
    const searchParams = new URLSearchParams();
    
    if (request.query) searchParams.append('query', request.query);
    if (request.status) request.status.forEach(s => searchParams.append('status', s));
    if (request.priority) request.priority.forEach(p => searchParams.append('priority', p));
    if (request.assignee_id !== undefined) searchParams.append('assignee_id', request.assignee_id.toString());
    if (request.parent_id !== undefined) searchParams.append('parent_id', request.parent_id.toString());
    if (request.has_subtasks !== undefined) searchParams.append('has_subtasks', request.has_subtasks.toString());
    if (request.has_dependencies !== undefined) searchParams.append('has_dependencies', request.has_dependencies.toString());
    if (request.due_date_from) searchParams.append('due_date_from', request.due_date_from);
    if (request.due_date_to) searchParams.append('due_date_to', request.due_date_to);
    if (request.created_from) searchParams.append('created_from', request.created_from);
    if (request.created_to) searchParams.append('created_to', request.created_to);
    if (request.skip !== undefined) searchParams.append('skip', request.skip.toString());
    if (request.limit !== undefined) searchParams.append('limit', request.limit.toString());
    if (request.sort_by) searchParams.append('sort_by', request.sort_by);
    if (request.sort_order) searchParams.append('sort_order', request.sort_order);
    
    const response = await api.get(`/projects/${projectId}/tasks/search?${searchParams.toString()}`);
    return response.data;
  },

  // Get task statistics
  async getTaskStats(): Promise<TaskStats> {
    const response = await api.get('/tasks/stats');
    return response.data;
  },

  // Get project task statistics
  async getProjectTaskStats(projectId: number): Promise<TaskStats> {
    const response = await api.get(`/projects/${projectId}/tasks/stats`);
    return response.data;
  },

  // Get task logs
  async getTaskLogs(taskId: number, params?: {
    skip?: number;
    limit?: number;
  }): Promise<TaskLog[]> {
    const searchParams = new URLSearchParams();
    
    if (params?.skip !== undefined) searchParams.append('skip', params.skip.toString());
    if (params?.limit !== undefined) searchParams.append('limit', params.limit.toString());
    
    const response = await api.get(`/tasks/${taskId}/logs?${searchParams.toString()}`);
    return response.data;
  },

  // Batch update tasks
  async batchUpdateTasks(data: TaskBatchUpdate): Promise<Task[]> {
    const response = await api.post('/tasks/batch/update', data);
    return response.data;
  },

  // Batch update task status
  async batchUpdateStatus(data: TaskBatchStatusUpdate): Promise<Task[]> {
    const response = await api.post('/tasks/batch/status', data);
    return response.data;
  },

  // Add task dependency
  async addDependency(taskId: number, dependsOnId: number): Promise<void> {
    await api.post(`/tasks/${taskId}/dependencies/${dependsOnId}`);
  },

  // Remove task dependency
  async removeDependency(taskId: number, dependsOnId: number): Promise<void> {
    await api.delete(`/tasks/${taskId}/dependencies/${dependsOnId}`);
  },

  // Helper methods
  async getTasksByStatus(status: TaskStatus, projectId?: number): Promise<TaskListItem[]> {
    if (projectId) {
      return this.getProjectTasks(projectId, {});
    }
    return this.getTasks({ status: [status] });
  },

  async getSubtasks(parentId: number, projectId?: number): Promise<TaskListItem[]> {
    if (projectId) {
      return this.getProjectTasks(projectId, { parent_id: parentId });
    }
    return this.getTasks({ parent_id: parentId });
  },

  async duplicateTask(taskId: number): Promise<Task> {
    const originalTask = await this.getTask(taskId);
    
    const duplicateData: TaskCreate = {
      title: `${originalTask.title} (副本)`,
      description: originalTask.description,
      details: originalTask.details,
      test_strategy: originalTask.test_strategy,
      priority: originalTask.priority,
      estimated_hours: originalTask.estimated_hours,
      project_id: originalTask.project_id,
      parent_id: originalTask.parent_id,
    };
    
    return this.createTask(duplicateData);
  },

  async moveTask(taskId: number, newParentId?: number): Promise<Task> {
    return this.updateTask(taskId, { parent_id: newParentId });
  },

  async getTaskCount(projectId?: number, status?: TaskStatus): Promise<number> {
    const tasks = await this.getTasks({ 
      project_id: projectId, 
      status: status ? [status] : undefined 
    });
    return tasks.length;
  },
};